### ISSUE TITLE
Exception when combining Omnifaces and PrimeFaces LazyDataModel on Payara Micro

### ISSUE THREAD

**Issue by alexanderder** [payara/Payara#884]
I get an Exception when adding the OmniFaces maven dependency to a Project, run on Payara Micro and work with PrimeFaces LazyDataModel. (There is a stackoverflow question, already: http://stackoverflow.com/questions/38087226)

```
[2016-06-28T17:19:09.280+0200] [Payara Micro 4.1] [SEVERE] [AS-WEB-CORE-00037] [
javax.enterprise.web.core] [tid: _ThreadID=17 _ThreadName=http-listener(7)] [tim
eMillis: 1467127149280] [levelValue: 1000] [[
  An exception or error occurred in the container during the request processing
java.lang.ClassCastException: cannot assign instance of java.lang.String to fiel
d org.primefaces.model.LazyDataModel.data of type java.util.List in instance of
org.primefaces.showcase.view.data.datatable.LazyCarDataModel
        at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStre
amClass.java:2083)
        at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:12
61)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:19
95)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1913)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:19
89)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1913)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
        at java.util.HashMap.readObject(HashMap.java:1155)
        at sun.reflect.GeneratedMethodAccessor48.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAcces
sorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:101
7)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1891)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
        at java.util.HashMap.readObject(HashMap.java:1155)
        at sun.reflect.GeneratedMethodAccessor48.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAcces
sorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:101
7)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1891)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:19
89)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1913)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
        at org.apache.catalina.session.StandardSession.readRemainingObject(Stand
ardSession.java:2079)
        at org.apache.catalina.session.StandardSession.readObject(StandardSessio
n.java:1987)
        at sun.reflect.GeneratedMethodAccessor49.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAcces
sorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:101
7)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1891)

        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1
796)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1348)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370)
        at org.apache.catalina.session.StandardSession.deserialize(StandardSessi
on.java:1274)
        at org.apache.catalina.session.StoreBase.readSession(StoreBase.java:296)

        at org.glassfish.web.ha.session.management.ReplicationStore.getSession(R
eplicationStore.java:567)
        at org.glassfish.web.ha.session.management.ReplicationStore.getSession(R
eplicationStore.java:519)
        at org.glassfish.web.ha.session.management.ReplicationStore.loadFromBack
ingStore(ReplicationStore.java:433)
        at org.glassfish.web.ha.session.management.ReplicationStore.load(Replica
tionStore.java:416)
        at org.apache.catalina.session.PersistentManagerBase.doSwapIn(Persistent
ManagerBase.java:1215)
        at org.apache.catalina.session.PersistentManagerBase.swapIn(PersistentMa
nagerBase.java:1175)
        at org.glassfish.web.ha.session.management.ReplicationManagerBase.findSe
ssion(ReplicationManagerBase.java:161)
        at org.apache.catalina.connector.Request.isRequestedSessionIdValid(Reque
st.java:2996)
        at org.apache.catalina.connector.Request.parseSessionCookiesId(Request.j
ava:4113)
        at org.apache.catalina.connector.CoyoteAdapter.postParseRequest(CoyoteAd
apter.java:672)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.j
ava:354)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.jav
a:283)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallab
le.call(ContainerMapper.java:459)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(Container
Mapper.java:167)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.
java:206)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.ja
va:180)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpSer
verFilter.java:235)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(Executor
Resolver.java:119)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(De
faultFilterChain.java:284)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart
(DefaultFilterChain.java:201)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultF
ilterChain.java:133)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultF
ilterChain.java:112)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.jav
a:77)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNI
OTransport.java:526)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(Abstr
actIOStrategy.java:112)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerTh
readIOStrategy.java:117)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(Wo
rkerThreadIOStrategy.java:56)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadR
unnable.run(WorkerThreadIOStrategy.java:137)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(Abs
tractThreadPool.java:591)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(Abstra
ctThreadPool.java:571)
        at java.lang.Thread.run(Thread.java:724)
]]
```

I can reproduce the Exception with the PrimeFaces 6.0 Showcase, so it looks like the cause is not within my code.

This is how to reproduce with PrimeFaces showcase:

(1) Download PrimeFaces 6.0 showcase sources from github.

(2) Add an OnmiFaces dependency to pom.xml:

```
<dependency>
    <groupId>org.omnifaces</groupId>
    <artifactId>omnifaces</artifactId>
    <version>2.3</version>
</dependency>
```

(3) Add the Maven code for Payara Micro Uber Jar creation from http://blog.payara.fish/creating-uber-jar-with-payara-micro to pom.xml

(4) Build with `mvn package` and then start with` java -jar showcase-6.0.jar`

(5) Now start the showcase on localhost and open the DataTable Lazy Loading example and you should get the exception.

The exeption does not occur, when the OmniFaces dependency is removed. And it does not occur on Payara Application Server, but only on Payara Micro.

(Maybe there is something wrong with Serialization ???)

**Comment by BalusC** [comment]
It appears that Payara Micro assumes the entire WAR as distributable when only a `web-fragment.xml` has this flag set. This is not correct. It should first check if WAR's own `web.xml` has this flag set, and then check if all of `web-fragment.xml` files also have this flag set. If this is absent in at least one of `web-fragment.xml`, then it should merely issue a warning that the WAR couldn't be set distributable.

**Comment by lprimak** [comment]
@BalusC This has been fixed by https://github.com/payara/Payara/issues/688 and should be available in the next version of Payara.

Still doesn't explain why List is assigned to a String

**Comment by smillidge** [comment]
Also Payara Micro runs clustered by default and deploys an application in a clustered mode. You can run Payara Micro with the --noCluster option and see if that makes a difference.

**Comment by simeo-reig** [comment]
It works with noCluster, thanks!

**Comment by lprimak** [comment]
When released, #1290 will fix this and generally improve session replication in Payara
